crypto/x509/x509_vfy.c

```diff
@@ -796,57 +796,53 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
     }
 
     /*
-     * Check trusted certificates in chain at depth num_untrusted and up.
-     * Note, that depths 0..num_untrusted-1 may also contain trusted
+     * Check newly added trust store element(s) in chain at depth num_untrusted
+     * and up.  Note, that depths 0..num_untrusted-1 may also contain trusted
      * certificates, but the caller is expected to have already checked those,
      * and wants to incrementally check just any added since.
      */
     for (i = num_untrusted; i < num; i++) {
         x = sk_X509_value(ctx->chain, i);
-        trust = X509_check_trust(x, ctx->param->trust, 0);
-        /* If explicitly trusted return trusted */
+        trust = X509_check_trust(x, ctx->param->trust, flags);
+        /*
+         * The chain is PKIX trusted if the cert is explicitly trusted or
+         * is implicitly trusted (self-signed or partial chains are allowed).
+         */
         if (trust == X509_TRUST_TRUSTED)
             goto trusted;
         if (trust == X509_TRUST_REJECTED)
             goto rejected;
     }
 
     /*
-     * If we are looking at a trusted certificate, and accept partial chains,
-     * the chain is PKIX trusted.
+     * Check the current cert for a direct trust store match. It is PKIX trusted
+     * if it is explicitly trusted or self-signed or partial chains are allowed.
      */
-    if (num_untrusted < num) {
-        if (ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN)
-            goto trusted;
-        return X509_TRUST_UNTRUSTED;
-    }
-
-    if (num_untrusted == num && ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN) {
-        /*
-         * Last-resort call with no new trusted certificates, check the leaf
-         * for a direct trust store match.
-         */
-        i = 0;
-        x = sk_X509_value(ctx->chain, i);
-        mx = lookup_cert_match(ctx, x);
-        if (mx == NULL)
-            return X509_TRUST_UNTRUSTED;
-
+    x = sk_X509_value(ctx->chain, num_untrusted - 1);
+    /*
+     * We check for an exact match to avoid
+     * possible impersonation via key substitution etc.
+     */
+    if (num_untrusted == num
+        && (mx = lookup_cert_match(ctx, x)) != NULL) {
```

Comment:
As I wrote in the description of this PR:
> it makes little sense to try finding an issuer of the current end of the chain if this end can be found immediately in the trust store (and is actually trusted).

Reply:
Other than the EE cert, we always know the provenance of each subsequent certificate as we're building the chain, because we added it, either by getting it from the trust store or from the untrusted store (often peer-provided chain). This is tracked by virtue of the "num_untrusted" variable.  Therefore, I still fail to see the point (which you're failing to explain precisely in a non hand waving way) of applying the EE-special trade-in for a trust-store equivalent at any other depth in the chain.

You just have to explain this much better.

Reply:
Good point that for each cert in the chain being constructed, except for the target cert (which BTW does not need to be an EE cert) we already know whether its is from the trust store, and in fact we already call `check_trust()` on the current cert if it is from the trust store, and stop the loop if its is either trusted or rejected!
This was not clear to me because the whole chain building loop is so spread-out and complicated with detail (which partly could and better should be abstracted away).
So thanks to your comment I meanwhile see that for the certs after the target cert the new check is actually redundant!

Reply:
> Good point that for each cert in the chain being constructed, except for the target cert (which BTW does not need to be an EE cert) we already know whether its is from the trust store, and in fact we already call `check_trust()` on the current cert if it is from the trust store, and stop the loop if its is either trusted or rejected!
> This was not clear to me because the whole chain building loop is so spread-out and complicated with detail (which partly could and better should be abstracted away).
> So thanks to your comment I meanwhile see that for the certs after the target cert the new check is actually redundant!

We're making progress!  If you limit the new check for just depth 0, with a suitable comment, my objections may be resolved.  You're then just handling EE direct match for partial chain early, skipping the chain construction. That's probably OK, if that's what you're after.  I can support that change, building a more complete chain seems redundant (the caller asked for partial chains, *and* has the EE cert in the store, could they possible indicate intent any more clearly???)

Reply:
Done this, as soon as possible, as follows:
```
/*
 * Stop immediately if target cert is directly trusted/rejected.
 * This improves completeness and prevents inefficiency and spurious
 * errors, which can occur when it makes no sense to look for an issuer.
 */
if (num == 1 && (trust = check_trust(ctx, 1)) != X509_TRUST_UNTRUSTED)
    break;
```